### PR TITLE
Trim RPM package dependencies

### PR DESCRIPTION
**Pull request by martinpitt**
libvirt-daemon-kvm is is a meta-package that pulls in the qemu-kvm
meta-package; that has tons of desktop support dependencies, plus all of
the storage/network libvirt drivers. Most of them are not required to
run cockpit-machines, so don't impose them on servers.

Instead, only depend on the basic packages for interacting with existing
VMs, and add some extra conditional dependencies if the recommended
virt-install is present (to be able to create new VMs).

Promote libvirt-daemon-config-network from a test to a runtime
dependency, as without it, creating system VMs does not work.

See https://libvirt.org/kbase/rpm-deployment.html

----

Similar to https://github.com/cockpit-project/cockpituous/pull/473 . See [recent discussion](https://lists.fedoraproject.org/archives/list/server@lists.fedoraproject.org/thread/SZKLLRC6ZRBFRYF5KFTWH7Z6MZ2GRX7V/) on Fedora Server list.

 - Fix or find a workaround for detecting spice QXL capability: https://bugzilla.redhat.com/show_bug.cgi?id=2064594: PR #643

### CODE REVIEW

**Comment by martinpitt** [comment]
[some pixel diffs](https://logs.cockpit-project.org/logs/pull-622-20220315-064749-254704d7-fedora-35/log.html) due to container refresh, will handle these separately

**Comment by martinpitt** [comment]
Fedora rawhide failure is the known one (see https://github.com/cockpit-project/bots/pull/3067 and #624) but [c8s](https://artifacts.dev.testing-farm.io/bcb44a52-6960-4c42-a50f-20df82ee2143/) is significant. Many tests fail on

    error: unsupported configuration: domain configuration does not support video model 'qxl'

**Comment by martinpitt** [comment]
All green, [rawhide failure](https://artifacts.dev.testing-farm.io/3dd68ec2-998e-4dd8-a327-af270da0d21c/) is still the same (see #624). Ready for review.

File: packaging/cockpit-machines.spec.in [hunk 1]
```diff
@@ -35,7 +35,9 @@ Requires: cockpit-bridge >= 215
 %if 0%{?suse_version}
 Requires: libvirt-daemon-qemu
 %else
-Requires: libvirt-daemon-kvm
+Requires: libvirt-daemon-driver-qemu
```
Comment (KKoukiou) [inline packaging/cockpit-machines.spec.in:38]: i am afraid we need at least the network and the storage libvirt core packages here, otherwise virt-install will fail to create the default pool/network. (as seen now on the tasks container anaconda webui tests) did you try it out without?

Comment (martinpitt) [reply packaging/cockpit-machines.spec.in:38]: Hmm, `virt-install` ought to depend on these then, but doesn't. `virt-install` itself is only recommended, but we can do something like

    Requires: (libvirt-daemon-driver-storage-core if virt-install)

(and the same for network). WDYT?

Comment (martinpitt) [reply packaging/cockpit-machines.spec.in:38]: Done, thanks for pointing out!

File: packaging/cockpit-machines.spec.in [hunk 2]
```diff
@@ -35,7 +35,11 @@ Requires: cockpit-bridge >= 215
...
+Requires: (qemu-kvm-core or libvirt-daemon-kvm)
+Requires: (libvirt-daemon-driver-storage-core if virt-install)
+Requires: (libvirt-daemon-driver-interface if virt-install)
+Suggests: libvirt-daemon-kvm
```
Comment (sgallagher) [inline packaging/cockpit-machines.spec.in:42]: This line doesn't do (in Fedora) what you might think.

In Debian/Ubuntu, this would cue the package manager to provide a message to the user about other packages they might want to install. In Fedora, `Suggests:` is only used to predispose the dependency solver towards a particular package when more than one might provide the same functionality. If the dependency isn't resolved firmly in another way (alternative already installed or pulled in by something else in the transaction), the package from the `Suggests:` directive will be installed.

Comment (martinpitt) [reply packaging/cockpit-machines.spec.in:42]: Thanks @sgallagher -- then I figure I'd rather take it out again, otherwise we are back to installing unwanted clutter.

**Comment by martinpitt** [comment]
I tested this on Fedora 35:
 - `dnf remove qemu-common libvirt-libs`, now there are no QEMU/libvirt packages left, aside from two with similar names which are unrelated:
```
# rpm -qa|grep -E 'virt|qemu'
qemu-guest-agent-6.1.0-14.fc35.x86_64
virt-what-1.21-2.fc35.x86_64
```

- Install `dnf install /var/tmp/build/cockpit-machines-*.noarch.rpm`. This *still* picks `libvirt-daemon-kvm` even though it's not the first alternative dependency. I guess I'm spoiled by deb packages, where the first alternative is the prefered one. However, this is no worse than what we have right now for new installations.

- However, I can nudge it with `dnf install qemu-kvm-core /var/tmp/build/cockpit-machines-*.noarch.rpm` -- in other words, if a more minimal installation is already present, that OR dependency does its job.
```
Installing:
 cockpit-machines                             noarch           263.34.g3c152c3-1.fc35                            @commandline           760 k
 qemu-kvm-core                                x86_64           2:6.1.0-14.fc35                                   updates                 14 k
Installing dependencies:
 capstone                                     x86_64           4.0.2-7.fc35                                      fedora                 747 k
 cyrus-sasl                                   x86_64           2.1.27-14.fc35                                    updates                 70 k
 daxctl-libs                                  x86_64           72.1-1.fc35                                       updates                 40 k
 edk2-ovmf                                    noarch           20211126gitbb1bba3d7767-1.fc35                    updates                5.0 M
 genisoimage                                  x86_64           1.1.11-48.fc35                                    fedora                 324 k
 gnutls-dane                                  x86_64           3.7.2-3.fc35                                      updates                 31 k
 gnutls-utils                                 x86_64           3.7.2-3.fc35                                      updates                449 k
 hwdata                                       noarch           0.357-1.fc35                                      updates                1.5 M
 ipxe-roms-qemu                               noarch           20200823-7.git4bd064de.fc35                       fedora                 1.4 M
 libfdt                                       x86_64           1.6.1-2.fc35                                      fedora                  33 k
 libpmem                                      x86_64           1.11.0-3.fc35                                     fedora                 107 k
 librdmacm                                    x86_64           39.0-1.fc35                                       updates                 72 k
 libretls                                     x86_64           3.4.2-1.fc35                                      updates                 40 k
 libssh2                                      x86_64           1.10.0-1.fc35                                     fedora                 117 k
 libtpms                                      x86_64           0.9.2-0.20220106gite81d634c27.fc35.0              updates                391 k
 libusal                                      x86_64           1.1.11-48.fc35                                    fedora                 137 k
 libvirt-client                               x86_64           7.6.0-5.fc35                                      updates                328 k
 libvirt-daemon                               x86_64           7.6.0-5.fc35                                      updates                369 k
 libvirt-daemon-driver-interface              x86_64           7.6.0-5.fc35                                      updates                158 k
 libvirt-daemon-driver-qemu                   x86_64           7.6.0-5.fc35                                      updates                871 k
 libvirt-daemon-driver-storage-core           x86_64           7.6.0-5.fc35                                      updates                211 k
 libvirt-dbus                                 x86_64           1.4.0-4.fc35                                      fedora                  93 k
 libvirt-glib                                 x86_64           4.0.0-3.fc35                                      fedora                  43 k
 libvirt-libs                                 x86_64           7.6.0-5.fc35                                      updates                4.4 M
 libwsman1                                    x86_64           2.7.0-5.fc35                                      fedora                 138 k
 libxslt                                      x86_64           1.1.34-6.fc35                                     fedora                 241 k
 lzop                                         x86_64           1.04-7.fc35                                       fedora                  55 k
 ndctl-libs                                   x86_64           72.1-1.fc35                                       updates                 81 k
 netcat                                       x86_64           1.218-2.fc35                                      updates                 34 k
 numad                                        x86_64           0.5-35.20150602git.fc35                           fedora                  37 k
 osinfo-db                                    noarch           20220214-1.fc35                                   updates                256 k
 osinfo-db-tools                              x86_64           1.9.0-2.fc35                                      fedora                  66 k
 pixman                                       x86_64           0.40.0-4.fc35                                     fedora                 270 k
 python3-argcomplete                          noarch           1.12.3-3.fc35                                     fedora                  60 k
 python3-libvirt                              x86_64           7.6.0-1.fc35                                      fedora                 314 k
 qemu-common                                  x86_64           2:6.1.0-14.fc35                                   updates                481 k
 qemu-img                                     x86_64           2:6.1.0-14.fc35                                   updates                1.7 M
 qemu-system-x86-core                         x86_64           2:6.1.0-14.fc35                                   updates                7.0 M
 seabios-bin                                  noarch           1.15.0-1.fc35                                     updates                183 k
 seavgabios-bin                               noarch           1.15.0-1.fc35                                     updates                 38 k
 sgabios-bin                                  noarch           1:0.20180715git-7.fc35                            fedora                  10 k
 swtpm                                        x86_64           0.7.1-1.20220218git92a7035.fc35                   updates                 41 k
 swtpm-libs                                   x86_64           0.7.1-1.20220218git92a7035.fc35                   updates                 48 k
 swtpm-tools                                  x86_64           0.7.1-1.20220218git92a7035.fc35                   updates                117 k
 virt-manager-common                          noarch           3.2.0-4.fc35                                      fedora                 953 k
 xen-libs                                     x86_64           4.15.2-1.fc35                                     updates                637 k
 xen-licenses                                 x86_64           4.15.2-1.fc35                                     updates                 49 k
Installing weak dependencies:
 libosinfo                                    x86_64           1.9.0-2.fc35                                      fedora                 295 k
 virt-install                                 noarch           3.2.0-4.fc35                                      fedora                  39 k
```

This is reasonable. I can log into Cockpit, and look at existing VMs. However, creating a new VM with "download an OS" or even a cloud image fails because of missing "qxl" support. Why is virt-install so insistent on adding a graphics card even if nobody asked for it? This isn't necessary at all for almost all use cases. `qemu-device-display-qxl` is one of the packages which have a ton of dependencies, like wayland, mesa, lots of libX*, libogg/libtheora (!!!), and gstreamer. None of that belongs onto a server really.

Can we tell virt-install to be a bit more clever about this, or do the detection ourselves? If qxl is available, then add it, but otherwise only add VNC?

**Comment by martinpitt** [comment]
Removing release-blocker. This is a whole lot more yak shaving than I initially anticipated.

**Comment by martinpitt** [comment]
> Can we tell virt-install to be a bit more clever about this, or do the detection ourselves? If qxl is available, then add it, but otherwise only add VNC?

Our `install_machine.py` actually tries to be clever about it, but `virsh domcapabilities` is lying  -- it still claims `spice` graphics support even if `qemu-device-display-qxl` is not installed. I'll file a bugzilla about it with a reproducer.

**Comment by martinpitt** [comment]
I adjusted a few more dependencies, and used the `Suggests:` trick to nudge dnf into the right direction, so that a default install is now reasonable:

```
# dnf install /var/tmp/cockpit-machines-263.35.g47a9566-1.fc35.noarch.rpm
Updating Subscription Management repositories.
Unable to read consumer identity

This system is not registered with an entitlement server. You can use subscription-manager to register.

Last metadata expiration check: 0:30:06 ago on Wed Mar 16 07:58:21 2022.
Dependencies resolved.
==============================================================================================================================================
 Package                                      Architecture     Version                                           Repository              Size
==============================================================================================================================================
Installing:
 cockpit-machines                             noarch           263.35.g47a9566-1.fc35                            @commandline           760 k
Installing dependencies:
 capstone                                     x86_64           4.0.2-7.fc35                                      fedora                 747 k
 cyrus-sasl                                   x86_64           2.1.27-14.fc35                                    updates                 70 k
 daxctl-libs                                  x86_64           72.1-1.fc35                                       updates                 40 k
 edk2-ovmf                                    noarch           20211126gitbb1bba3d7767-1.fc35                    updates                5.0 M
 genisoimage                                  x86_64           1.1.11-48.fc35                                    fedora                 324 k
 gnutls-dane                                  x86_64           3.7.2-3.fc35                                      updates                 31 k
 gnutls-utils                                 x86_64           3.7.2-3.fc35                                      updates                449 k
 hwdata                                       noarch           0.357-1.fc35                                      updates                1.5 M
 ipxe-roms-qemu                               noarch           20200823-7.git4bd064de.fc35                       fedora                 1.4 M
 libfdt                                       x86_64           1.6.1-2.fc35                                      fedora                  33 k
 libpmem                                      x86_64           1.11.0-3.fc35                                     fedora                 107 k
 librdmacm                                    x86_64           39.0-1.fc35                                       updates                 72 k
 libretls                                     x86_64           3.4.2-1.fc35                                      updates                 40 k
 libssh2                                      x86_64           1.10.0-1.fc35                                     fedora                 117 k
 libtpms                                      x86_64           0.9.2-0.20220106gite81d634c27.fc35.0              updates                391 k
 libusal                                      x86_64           1.1.11-48.fc35                                    fedora                 137 k
 libvirt-client                               x86_64           7.6.0-5.fc35                                      updates                328 k
 libvirt-daemon                               x86_64           7.6.0-5.fc35                                      updates                369 k
 libvirt-daemon-config-network                x86_64           7.6.0-5.fc35                                      updates                 14 k
 libvirt-daemon-driver-interface              x86_64           7.6.0-5.fc35                                      updates                158 k
 libvirt-daemon-driver-network                x86_64           7.6.0-5.fc35                                      updates                195 k
 libvirt-daemon-driver-qemu                   x86_64           7.6.0-5.fc35                                      updates                871 k
 libvirt-daemon-driver-storage-core           x86_64           7.6.0-5.fc35                                      updates                211 k
 libvirt-dbus                                 x86_64           1.4.0-4.fc35                                      fedora                  93 k
 libvirt-glib                                 x86_64           4.0.0-3.fc35                                      fedora                  43 k
 libvirt-libs                                 x86_64           7.6.0-5.fc35                                      updates                4.4 M
 libwsman1                                    x86_64           2.7.0-5.fc35                                      fedora                 138 k
 libxslt                                      x86_64           1.1.34-6.fc35                                     fedora                 241 k
 lzop                                         x86_64           1.04-7.fc35                                       fedora                  55 k
 ndctl-libs                                   x86_64           72.1-1.fc35                                       updates                 81 k
 netcat                                       x86_64           1.218-2.fc35                                      updates                 34 k
 numad                                        x86_64           0.5-35.20150602git.fc35                           fedora                  37 k
 osinfo-db                                    noarch           20220214-1.fc35                                   updates                256 k
 osinfo-db-tools                              x86_64           1.9.0-2.fc35                                      fedora                  66 k
 pixman                                       x86_64           0.40.0-4.fc35                                     fedora                 270 k
 python3-argcomplete                          noarch           1.12.3-3.fc35                                     fedora                  60 k
 python3-libvirt                              x86_64           7.6.0-1.fc35                                      fedora                 314 k
 qemu-common                                  x86_64           2:6.1.0-14.fc35                                   updates                481 k
 qemu-img                                     x86_64           2:6.1.0-14.fc35                                   updates                1.7 M
 qemu-kvm-core                                x86_64           2:6.1.0-14.fc35                                   updates                 14 k
 qemu-system-x86-core                         x86_64           2:6.1.0-14.fc35                                   updates                7.0 M
 radvd                                        x86_64           2.19-4.fc35                                       fedora                  87 k
 seabios-bin                                  noarch           1.15.0-1.fc35                                     updates                183 k
 seavgabios-bin                               noarch           1.15.0-1.fc35                                     updates                 38 k
 sgabios-bin                                  noarch           1:0.20180715git-7.fc35                            fedora                  10 k
 swtpm                                        x86_64           0.7.1-1.20220218git92a7035.fc35                   updates                 41 k
 swtpm-libs                                   x86_64           0.7.1-1.20220218git92a7035.fc35                   updates                 48 k
 swtpm-tools                                  x86_64           0.7.1-1.20220218git92a7035.fc35                   updates                117 k
 virt-manager-common                          noarch           3.2.0-4.fc35                                      fedora                 953 k
 xen-libs                                     x86_64           4.15.2-1.fc35                                     updates                637 k
 xen-licenses                                 x86_64           4.15.2-1.fc35                                     updates                 49 k
Installing weak dependencies:
 libosinfo                                    x86_64           1.9.0-2.fc35                                      fedora                 295 k
 virt-install                                 noarch           3.2.0-4.fc35                                      fedora                  39 k
```

That still runs into the qxl issue, but once that is fixed/worked around, it works. I hacked our `install_script.py` to drop `--graphics spice`, and all the earlier follow-up errors about missing default network etc. are gone:

```
python3 /tmp/install_machine.py '{"connectionName":"system","memorySize":128,"os":"cirros0.5.2","rootPassword":"","source":"/tmp/cirros.qcow2","sourceType":"cloud","startVm":true,"storagePool":"NewVolume","storageSize":0.048828125,"storageVolume":"","type":"create","unattended":false,"userLogin":"","userPassword":"","vmName":"t1"}'
```

And I see the CirrOS instance booting happily in c-machines.

**Comment by martinpitt** [comment]
I filed https://bugzilla.redhat.com/show_bug.cgi?id=2064594 about the spice detection. I didn't find a good workaround for this yet, some help here would be appreciated.

**Comment by sgallagher** [comment]
> * Install `dnf install /var/tmp/build/cockpit-machines-*.noarch.rpm`. This _still_ picks `libvirt-daemon-kvm` even though it's not the first alternative dependency. I guess I'm spoiled by deb packages, where the first alternative is the prefered one. However, this is no worse than what we have right now for new installations.

This is exactly the case that `Suggests:` is for. If there's a better alternative, have cockpit-machines `Suggests:` it.

**Comment by martinpitt** [comment]
@sgallagher : Right! That's what I did now. There's some further test fallout, but before I look at this we need some solution for the "detect spice capability" anyway, until then I put this on hold.

**Comment by martinpitt** [comment]
@sgallagher : This still pulls in a lot of things. I am particularly worried about new systemd services like device-mapper, glusterfs, iscsi, and fundamental functionality like systemd-networkd. But it also won't help with the spice detection issue. The proposed dependencies here are more complex admittedly (and I'd love to simplify them), but libvirtd seems too heavy.

This is probably also a philosophical matter: Should installing cockpit-machines automatically give you all the functionality that it can do (such as iSCSI) or not impose anything onto your system which you are not already using? I'm rather firmly leaning towards the latter.

**Comment by sgallagher** [comment]
@martinpitt In general, I agree with you *if* `cockpit-machines` could at least provide the option to install those missing pieces on-demand. Without that (or at least directions on how to do it yourself), I think people will not get what they are expecting.

**Review by sgallagher — APPROVED** [review]
LGTM :+1:

File: test/browser/main.fmf [hunk 3]
```diff
@@ -14,7 +14,8 @@ require:
   - python3-yaml
   # required by tests
   - firewalld
-  - libvirt-daemon-config-network
+  - libvirt-daemon-driver-storage-iscsi
+  - libvirt-daemon-driver-storage-logical
```
Comment (skobyda) [inline test/browser/main.fmf:18]: I understand other changes where we replace general package with more specific ones, but what is the motivation for this change?
`libvirt-daemon-config-network` and `libvirt-daemon-driver-storage-*` seem to serve different purposes. 
Is libvirt-daemon-config-network redundant? are libvirt-daemon-driver-storage-* suddenly needed?

Comment (martinpitt) [reply test/browser/main.fmf:18]: The removal and addition are independent from each other, they just happened to be at the same place. As I wrote in the commit message, libvirt-daemon-config-network got moved to the .spec:

    Requires: (libvirt-daemon-config-network if virt-install)

as I noticed that creating a system VM fails on the missing default network. So it really shouldn't be a test-only dependency. I think I'll move this into a separate commit to explain it better.

The tests cover iscsi and LVM storage pools. They were previously pulled in via libvirt-daemon-kvm, but I replaced the storage parts with libvirt-daemon-driver-storage-core. That will just give you basic directories, but avoids pulling in iSCSI or LVM (which include udev rules and running services!) on machines which don't need/want it.

Comment (martinpitt) [reply test/browser/main.fmf:18]: I split out the config-network commit, hopefully that clarifies it a bit?

Comment (skobyda) [reply test/browser/main.fmf:18]: Ah yeah thanks, I just thought the removal of the network package and the addition of the storage package are related to each other, and that confused me

**Review by skobyda — APPROVED** [review]

**Comment by martinpitt** [comment]
c8s is having a bad day (broken archive *again*), unrelated.

I saw [this rawhide failure](https://artifacts.dev.testing-farm.io/5c5e74c9-1b02-4fcc-bfc6-596891fe5c30/) recently in other PRs, it's a new flake. But this was green before the commit split (with net-zero effect), so I land this now and look at that in #660.
